pyggi/base/program.py

```diff
@@ -267,14 +269,16 @@ def exec_cmd(self, cmd, timeout=15):
         sprocess = subprocess.Popen(
             shlex.split(cmd),
             stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE)
+            stderr=subprocess.PIPE,
+            preexec_fn=os.setsid)
         try:
             start = time.time()
             stdout, stderr = sprocess.communicate(timeout=timeout)
             end = time.time()
             return (sprocess.returncode, stdout.decode("ascii"), stderr.decode("ascii"), end-start)
         except subprocess.TimeoutExpired:
-            sprocess.kill()
+            os.killpg(os.getpgid(sprocess.pid), signal.SIGKILL)
+            _, _ = sprocess.communicate()
```

Comment:
Aymeric, could you explain why `communicate` is called here?

Reply:
Tricky question.
Short answer: it's there for unrelated reasons and can go to another commit.

Documentation: https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate
I think originally I added that because of that passage: _"in order to cleanup properly a well-behaved application should kill the child process and finish communication"_; it actually predates the change from `kill` to `killpg`.
According to the doc this call does three things: (1) flushing stdout/stderr, (2) waiting for the process to be terminated, and (3) setting the `returncode` attribute.
I concede that (1) seems relatively unimportant.
(2) however prevents PyGGI to spawn cohorts of zombies when something goes very wrong (at worse it spawns a single one and yields).
As for (3) I will need it for another commit (3f3ee1fb7ba1d34428b8bcb2ba036e76b3c09428 ; but this one require special attention) so it can be merged there instead.

Just let me know and I will rebase it away.